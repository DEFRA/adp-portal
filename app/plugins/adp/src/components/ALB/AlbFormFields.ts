export const albFormFields = [
    {
      label: 'Name',
      name: 'name',
      helperText:
        'This must be unique - use letters, numbers, or separators such as "_", "-"',
      validations: {
        required: true,
        pattern: {
          value: /^([a-zA-Z0-9 ]+[-_. ]?)*[a-zA-Z0-9]+$/,
          message:
            'Invalid ALB name format. Use letters, numbers, or "-", "_", "." as separators.',
        },
      },
    },
    {
      label: 'Short Name',
      name: 'short_name',
      helperText: 'Optional - a short form name to identify the body',
    },
    {
      label: 'ALB Description',
      name: 'description',
      helperText: 'Max 200 Chars',
      validations: {
        required: true,
        maxLength: 200,
      },
      multiline: true,
      maxRows: 4,
    },
  ];