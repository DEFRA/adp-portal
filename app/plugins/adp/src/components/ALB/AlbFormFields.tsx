import React from 'react';
import { DisabledFields, FormTextField, formRules } from '../../utils';
import { UseFormReturn } from 'react-hook-form';

export type AlbFields = {
  title: string;
  alias: string;
  url: string;
  description: string;
};

export type AlbFormFieldsProps = Readonly<
  UseFormReturn<AlbFields> & {
    disabled?: DisabledFields<AlbFields>;
  }
>;
export const emptyForm = Object.freeze<AlbFields>({
  alias: '',
  description: '',
  title: '',
  url: '',
});

export function AlbFormFields({
  control,
  formState: { errors },
  disabled,
}: AlbFormFieldsProps) {
  let i = 0;
  return (
    <>
      <FormTextField
        control={control}
        errors={errors}
        index={i++}
        name="title"
        label="Name"
        helperText="This must be unique"
        disabled={disabled}
        rules={{
          ...formRules.required,
          ...formRules.maxLength(100),
        }}
      />

      <FormTextField
        control={control}
        errors={errors}
        index={i++}
        name="alias"
        label="Alias"
        helperText="Are there any other names this Arms Length Body is also known as?"
        disabled={disabled}
      />

      <FormTextField
        control={control}
        errors={errors}
        index={i++}
        name="url"
        label="Website"
        helperText="A link to a website"
        disabled={disabled}
        rules={{
          ...formRules.pattern(/^https?:\/\//, 'Invalid url format'),
        }}
      />

      <FormTextField
        control={control}
        errors={errors}
        index={i++}
        name="description"
        label="Description"
        helperText="Max 200 characters"
        disabled={disabled}
        maxRows={4}
        rules={{
          ...formRules.required,
          ...formRules.maxLength(200),
        }}
      />
    </>
  );
}
