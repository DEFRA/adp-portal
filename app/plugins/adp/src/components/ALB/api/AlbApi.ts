import {
  ArmsLengthBody,
  CreateArmsLengthBodyRequest,
  UpdateArmsLengthBodyRequest,
} from '@internal/plugin-adp-common';
import { createApiRef } from '@backstage/core-plugin-api';

export const armsLengthBodyApiRef = createApiRef<ArmsLengthBodyApi>({
  id: 'plugin.adp.armslengthbodyapi',
});

export interface ArmsLengthBodyApi {
  getArmsLengthBodies(): Promise<ArmsLengthBody[]>;
  updateArmsLengthBody(
    data: UpdateArmsLengthBodyRequest,
  ): Promise<ArmsLengthBody[]>;
  createArmsLengthBody(
    data: CreateArmsLengthBodyRequest,
  ): Promise<ArmsLengthBody[]>;
  getArmsLengthBodyNames(): Promise<Record<string, string>>;
}
