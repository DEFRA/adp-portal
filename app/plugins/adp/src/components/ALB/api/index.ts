export { ArmsLengthBodyClient } from './AlbClient'
export type {ArmsLengthBodyApi} from './AlbApi'
