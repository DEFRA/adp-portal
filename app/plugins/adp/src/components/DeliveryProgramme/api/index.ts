export { DeliveryProgrammeClient } from './DeliveryProgrammeClient';
export type { DeliveryProgrammeApi } from './DeliveryProgrammeApi';
