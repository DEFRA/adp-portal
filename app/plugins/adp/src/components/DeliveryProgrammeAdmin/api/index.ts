export { deliveryProgrammeAdminApiRef } from './DeliveryProgrammeAdminApi';
export { DeliveryProgrammeAdminApiClient } from './DeliveryProgrammeAdminClient';
