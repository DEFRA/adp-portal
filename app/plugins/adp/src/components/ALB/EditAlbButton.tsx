import React, { useState } from 'react';
import { Button } from '@material-ui/core';
import { alertApiRef, useApi } from '@backstage/core-plugin-api';
import type { AlbFields} from './AlbFormFields';
import { AlbFormFields } from './AlbFormFields';
import { usePermission } from '@backstage/plugin-permission-react';
import type {
  ArmsLengthBody} from '@internal/plugin-adp-common';
import {
  adpProgrammmeCreatePermission,
} from '@internal/plugin-adp-common';
import type { SubmitResult} from '../../utils';
import { DialogForm, readValidationError } from '../../utils';
import { armsLengthBodyApiRef } from './api';

export type EditAlbButtonProps = Readonly<
  Omit<Parameters<typeof Button>[0], 'onClick'> & {
    armsLengthBody: ArmsLengthBody;
    onEdited?: () => void;
  }
>;

export function EditAlbButton({
  onEdited,
  armsLengthBody,
  children,
  ...buttonProps
}: EditAlbButtonProps) {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const alertApi = useApi(alertApiRef);
  const client = useApi(armsLengthBodyApiRef);

  const { allowed: allowedToCreateAlb } = usePermission({
    permission: adpProgrammmeCreatePermission,
  });
  if (!allowedToCreateAlb) return null;

  async function handleSubmit(
    fields: AlbFields,
  ): Promise<SubmitResult<AlbFields>> {
    try {
      await client.updateArmsLengthBody({
        ...fields,
        id: armsLengthBody.id,
      });
    } catch (e: any) {
      return readValidationError(e);
    }
    alertApi.post({
      message: 'ALB edited successfully.',
      severity: 'success',
      display: 'transient',
    });
    return { type: 'success' };
  }

  return (
    <>
      <Button
        data-testid="alb-edit-button"
        {...buttonProps}
        onClick={() => setIsModalOpen(true)}
      >
        {children}
      </Button>
      {isModalOpen && (
        <DialogForm
          renderFields={AlbFormFields}
          defaultValues={armsLengthBody}
          completed={form => {
            setIsModalOpen(false);
            if (form) onEdited?.();
          }}
          title={`Update ALB ${armsLengthBody.title}`}
          confirm="Update"
          submit={handleSubmit}
        />
      )}
    </>
  );
}
