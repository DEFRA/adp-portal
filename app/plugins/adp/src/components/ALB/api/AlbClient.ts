import { ArmsLengthBodyApi } from './AlbApi';
import { ArmsLengthBody } from '@internal/plugin-adp-common';

import { DiscoveryApi, FetchApi } from '@backstage/core-plugin-api';
import { ResponseError } from '@backstage/errors';

export class ArmsLengthBodyClient implements ArmsLengthBodyApi {
  private discoveryApi: DiscoveryApi;
  private fetchApi: FetchApi;

  constructor(discoveryApi: DiscoveryApi, fetchApi: FetchApi) {
    this.discoveryApi = discoveryApi;
    this.fetchApi = fetchApi;
  }

  private async getApiUrl(): Promise<string> {
    return `${await this.discoveryApi.getBaseUrl('adp')}/armsLengthBody`;
  }

  async getArmsLengthBodies(): Promise<ArmsLengthBody[]> {
    try {
      const url = await this.getApiUrl();
      const response = await this.fetchApi.fetch(url);
      if (!response.ok) {
        throw await ResponseError.fromResponse(response);
      }
      return response.json();
    } catch (error) {
      throw new Error('Failed to fetch arms length bodies: ${error.message}');
    }
  }

  async createArmsLengthBody(data: any): Promise<ArmsLengthBody[]> {
    const url = await this.getApiUrl();

    const response = await this.fetchApi.fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    return response.json();
  }

  async updateArmsLengthBody(data: any): Promise<ArmsLengthBody[]> {
    const url = await this.getApiUrl();

    const response = await this.fetchApi.fetch(url, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    const updatedData: ArmsLengthBody[] = await response.json();
    return updatedData;
  }

  async getArmsLengthBodyNames(): Promise<Record<string, string>> { 
    try {
      const albNamesUrl = `${await this.discoveryApi.getBaseUrl('adp')}/armslengthbodynames`;
  
      const response = await this.fetchApi.fetch(albNamesUrl);
      if (!response.ok) {
        throw await ResponseError.fromResponse(response);
      }
      return response.json();
    } catch (error) {
      throw new Error(`Failed to fetch arms length bodies`); 
    }
  }
}

export type { ArmsLengthBodyApi };
