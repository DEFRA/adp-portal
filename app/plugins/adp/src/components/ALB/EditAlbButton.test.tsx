import type { AlertApi } from '@backstage/core-plugin-api';
import { alertApiRef } from '@backstage/core-plugin-api';
import React from 'react';
import type { ArmsLengthBodyApi } from './api';
import { armsLengthBodyApiRef } from './api';
import { render as testRender, waitFor } from '@testing-library/react';
import { TestApiProvider } from '@backstage/test-utils';
import type { EditAlbButtonProps } from './EditAlbButton';
import { EditAlbButton } from './EditAlbButton';
import userEvent from '@testing-library/user-event';
import type { AlbFields } from './AlbFormFields';
import { AlbFormFields } from './AlbFormFields';
import { act } from 'react-dom/test-utils';
import type {
  ArmsLengthBody,
  ValidationError as IValidationError,
} from '@internal/plugin-adp-common';
import { SnapshotFriendlyStylesProvider, ValidationError } from '../../utils';
import type * as PluginPermissionReactModule from '@backstage/plugin-permission-react';
import type * as DialogFormModule from '../../utils/DialogForm';

const usePermission: jest.MockedFn<
  typeof PluginPermissionReactModule.usePermission
> = jest.fn();
const DialogForm: jest.MockedFn<typeof DialogFormModule.DialogForm> = jest.fn();

const armsLengthBody: ArmsLengthBody = {
  created_at: new Date(0),
  creator: 'Me',
  description: 'My ALB',
  id: '00000000-0000-0000-0000-000000000001',
  name: 'my-cool-alb',
  owner: 'Someone',
  title: 'My cool ALB',
  updated_at: new Date(0),
  alias: 'CAB',
  url: 'https://test.com',
};

const fields: AlbFields = {
  alias: 'abc',
  description: 'def',
  title: 'ghi',
  url: 'jkl',
};

beforeEach(() => {
  jest.resetAllMocks();
});

describe('EditAlbButton', () => {
  it('Should not render when the user is not allowed to edit albs', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: false, loading: false });

    const { result } = await render({ content: 'My button', armsLengthBody });

    expect(result.baseElement).toMatchSnapshot();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
  });
  it('Should only render a button initially', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });

    const { result } = await render({ content: 'My button', armsLengthBody });

    expect(result.baseElement).toMatchSnapshot();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
  });
  it('Should render the dialog when the button is clicked', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });
    DialogForm.mockReturnValue(<span>This is a dialog!</span>);

    const { result } = await render({ content: 'My button', armsLengthBody });
    await userEvent.click(result.getByTestId('alb-edit-button'));

    expect(result.baseElement).toMatchSnapshot();
    expect(DialogForm.mock.calls).toHaveLength(1);
    const formProps = DialogForm.mock.calls[0][0];
    expect({
      title: formProps.title,
      open: formProps.open,
      renderFields: formProps.renderFields,
      confirm: formProps.confirm,
      cancel: formProps.cancel,
      defaultValues: formProps.defaultValues,
      disabled: formProps.disabled,
      validate: formProps.validate,
    }).toMatchObject({
      title: 'Update ALB My cool ALB',
      open: undefined,
      renderFields: AlbFormFields,
      confirm: 'Update',
      cancel: undefined,
      defaultValues: {
        description: armsLengthBody.description,
        title: armsLengthBody.title,
        url: armsLengthBody.url,
        alias: armsLengthBody.alias,
      },
      disabled: undefined,
      validate: undefined,
    });
    expect(formProps.submit).toBeDefined();
    expect(formProps.completed).toBeDefined();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
  });
  it('Should close the form when the form is cancelled.', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });
    DialogForm.mockReturnValue(<span>This is a dialog!</span>);

    const { result } = await render({ content: 'My button', armsLengthBody });
    await userEvent.click(result.getByTestId('alb-edit-button'));

    expect(result.baseElement).toMatchSnapshot('Before cancel');
    expect(DialogForm.mock.calls).toHaveLength(1);
    const formProps = DialogForm.mock.calls[0][0];
    act(() => formProps.completed(undefined));
    await waitFor(() =>
      expect(result.queryByText('This is a dialog!')).toBeNull(),
    );
    expect(result.baseElement).toMatchSnapshot('After cancel');
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
  });
  it('Should call onEditd when the form closes with a value.', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });
    DialogForm.mockReturnValue(<span>This is a dialog!</span>);
    const onEdited: jest.MockedFn<
      Exclude<EditAlbButtonProps['onEdited'], undefined>
    > = jest.fn();

    const { result } = await render({
      content: 'My button',
      armsLengthBody,
      onEdited,
    });
    await userEvent.click(result.getByTestId('alb-edit-button'));

    expect(result.baseElement).toMatchSnapshot('Before complete');
    expect(DialogForm.mock.calls).toHaveLength(1);
    expect(onEdited).not.toHaveBeenCalled();
    const formProps = DialogForm.mock.calls[0][0];
    act(() => formProps.completed(fields));
    await waitFor(() =>
      expect(result.queryByText('This is a dialog!')).toBeNull(),
    );
    expect(result.baseElement).toMatchSnapshot('After complete');
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(onEdited).toHaveBeenCalledTimes(1);
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
  });
  it('Should call the api when submitting.', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });
    DialogForm.mockReturnValue(<span>This is a dialog!</span>);

    const { result } = await render({ content: 'My button', armsLengthBody });
    await userEvent.click(result.getByTestId('alb-edit-button'));

    expect(result.baseElement).toMatchSnapshot();
    expect(DialogForm.mock.calls).toHaveLength(1);
    const formProps = DialogForm.mock.calls[0][0];
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();

    const submitResult = await formProps.submit(fields);
    expect(submitResult).toMatchObject({ type: 'success' });
    expect(mockArmsLengthBodyApi.updateArmsLengthBody.mock.calls).toMatchObject(
      [[fields]],
    );
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post.mock.calls).toMatchObject([
      [
        {
          message: 'ALB edited successfully.',
          severity: 'success',
          display: 'transient',
        },
      ],
    ]);
  });
  it('Should catch ValidationErrors when submitting.', async () => {
    const { mockAlertApi, mockArmsLengthBodyApi, render } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });
    DialogForm.mockReturnValue(<span>This is a dialog!</span>);
    const validationErrors: IValidationError[] = [
      {
        path: 'abc',
        error: {
          message: '123',
        },
      },
    ];
    mockArmsLengthBodyApi.updateArmsLengthBody.mockRejectedValueOnce(
      new ValidationError(validationErrors),
    );

    const { result } = await render({ content: 'My button', armsLengthBody });
    await userEvent.click(result.getByTestId('alb-edit-button'));

    expect(result.baseElement).toMatchSnapshot();
    expect(DialogForm.mock.calls).toHaveLength(1);
    const formProps = DialogForm.mock.calls[0][0];
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.updateArmsLengthBody).not.toHaveBeenCalled();

    const submitResult = await formProps.submit(fields);
    expect(submitResult).toMatchObject({
      type: 'validationError',
      errors: validationErrors,
    });
    expect(mockArmsLengthBodyApi.updateArmsLengthBody.mock.calls).toMatchObject(
      [[fields]],
    );
    expect(mockArmsLengthBodyApi.getArmsLengthBodies).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.getArmsLengthBodyNames).not.toHaveBeenCalled();
    expect(mockArmsLengthBodyApi.createArmsLengthBody).not.toHaveBeenCalled();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
  });
});

function setup() {
  const mockAlertApi: jest.Mocked<AlertApi> = {
    alert$: jest.fn(),
    post: jest.fn(),
  };
  const mockArmsLengthBodyApi: jest.Mocked<ArmsLengthBodyApi> = {
    createArmsLengthBody: jest.fn(),
    getArmsLengthBodies: jest.fn(),
    getArmsLengthBodyNames: jest.fn(),
    updateArmsLengthBody: jest.fn(),
  };

  return {
    mockAlertApi,
    mockArmsLengthBodyApi,
    async render(props: EditAlbButtonProps) {
      const result = testRender(
        <TestApiProvider
          apis={[
            [alertApiRef, mockAlertApi],
            [armsLengthBodyApiRef, mockArmsLengthBodyApi],
          ]}
        >
          <SnapshotFriendlyStylesProvider>
            <EditAlbButton {...props} />
          </SnapshotFriendlyStylesProvider>
        </TestApiProvider>,
      );
      await waitFor(() => expect(result.baseElement).not.toBeEmptyDOMElement());
      return { result };
    },
  };
}

jest.mock(
  '@backstage/plugin-permission-react',
  () =>
    ({
      get usePermission() {
        return usePermission;
      },
      get IdentityPermissionApi(): never {
        throw new Error('Not mocked');
      },
      get PermissionedRoute(): never {
        throw new Error('Not mocked');
      },
      get RequirePermission(): never {
        throw new Error('Not mocked');
      },
      get permissionApiRef(): never {
        throw new Error('Not mocked');
      },
    } satisfies typeof PluginPermissionReactModule),
);

jest.mock(
  '../../utils/DialogForm',
  () =>
    ({
      get DialogForm() {
        return DialogForm as typeof DialogFormModule.DialogForm;
      },
    } satisfies typeof DialogFormModule),
);
