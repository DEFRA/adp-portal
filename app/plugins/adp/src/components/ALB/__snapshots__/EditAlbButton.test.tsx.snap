// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditAlbButton Should call onEditd when the form closes with a value.: After complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
  </div>
</body>
`;

exports[`EditAlbButton Should call onEditd when the form closes with a value.: Before complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;

exports[`EditAlbButton Should call the api when submitting. 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;

exports[`EditAlbButton Should call the api when submitting.: After complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`EditAlbButton Should call the api when submitting.: Before complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`EditAlbButton Should catch ValidationErrors when submitting. 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;

exports[`EditAlbButton Should catch ValidationErrors when submitting. 2`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;

exports[`EditAlbButton Should close the form when the form is cancelled.: After cancel 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
  </div>
</body>
`;

exports[`EditAlbButton Should close the form when the form is cancelled.: Before cancel 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;

exports[`EditAlbButton Should not render when the user is not allowed to edit albs 1`] = `
<body>
  <div />
</body>
`;

exports[`EditAlbButton Should only render a button initially 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`EditAlbButton Should render the dialog when the button is clicked 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="My button"
      data-testid="alb-edit-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      This is a dialog!
    </span>
  </div>
</body>
`;
